### PR TITLE
Use ReflectionOnly as data contract serializer option for iOS/tvOS/MacCatalyst

### PR DESCRIPTION
**Pull request by MaximLipnin**
An attempt to address https://github.com/dotnet/runtime/issues/47114 by making the `System.Runtime.Serialization.DataContractSerializer.Option` property return `ReflectionOnly` value when `RuntimeFeature.IsDynamicCodeSupported` is true.

### CODE REVIEW

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs [hunk 1]
```diff
@@ -34,7 +34,17 @@ public sealed class DataContractSerializer : XmlObjectSerializer
         private ISerializationSurrogateProvider? _serializationSurrogateProvider;
         private bool _serializeReadOnlyTypes;
 
-        private static SerializationOption _option = IsReflectionBackupAllowed() ? SerializationOption.ReflectionAsBackup : SerializationOption.CodeGenOnly;
+        private static SerializationOption _option = SetSerializationOption();
+
+        private static SerializationOption SetSerializationOption()
```
Comment (filipnavara) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs:39]: nit: I'd think a better name could be `GetDefaultSerializationOption`.

**Review by filipnavara — APPROVED** [review]

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs [hunk 2]
```diff
@@ -34,7 +34,16 @@ public sealed class DataContractSerializer : XmlObjectSerializer
         private ISerializationSurrogateProvider? _serializationSurrogateProvider;
         private bool _serializeReadOnlyTypes;
 
-        private static SerializationOption _option = IsReflectionBackupAllowed() ? SerializationOption.ReflectionAsBackup : SerializationOption.CodeGenOnly;
+        private static SerializationOption _option = GetDefaultSerializationOption();
```
[the remark this answers is not in the document]
Comment (MaximLipnin) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Thanks, I followed the approach you suggested, but I'm not sure whether the `System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported` feature, which I've used, is correct here and I need to add something like
```
<IlcArg Include="--feature:System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported=true" />
```

[the remark this answers is not in the document]
Comment (jkotas) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Maybe you can just change the Option property to:
```
        internal static SerializationOption Option
        {
            get { return RuntimeFeature.IsDynamicCodeSupported ? _option : SerializationOption.ReflectionOnly; }
```
and not have any xml substitutions.

[the remark this answers is not in the document]
Comment (MaximLipnin) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContractSerializer.cs]: Done, thanks

**Review by steveisok — APPROVED** [review]
LGTM.  I know the iOS/tvOS/Catalyst test failures are unrelated.  Can you address them separately?

**Review by akoeplinger — APPROVED** [review]

**Comment by filipnavara** [comment]
Thanks!

**Pull request merged**